In [15]:
# Import the libraries we need for this project

import pandas as pd
import zipfile

# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

# Import the Logistic Regression model
from sklearn.linear_model import LogisticRegression

# Import evaluation metrics
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [16]:
# Open the ZIP file and load the Titanic dataset

with zipfile.ZipFile("archive.zip") as zip_ref:
    df = pd.read_csv(zip_ref.open("Titanic-Dataset.csv"))

# Display the first 5 rows
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [17]:
# Display basic information about the dataset
df.info()
# Check for missing values in each column
df.isnull().sum()


<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [18]:
# Remove columns that are not useful for prediction
df = df.drop(columns=["PassengerId", "Name", "Ticket", "Cabin"])
# Fill missing Age values with the average age
df["Age"] = df["Age"].fillna(df["Age"].mean())
# Fill missing Embarked values with the most common value
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])

In [19]:
# Machine learning models work only with numbers
# Convert categorical columns into numerical columns
df = pd.get_dummies(df, columns=["Sex", "Embarked"], drop_first=True)
# Display the cleaned dataset
df.head()

,Survived,Pclass,Age,SibSp,Parch,Fare,Sex_male,Embarked_Q,Embarked_S
0,0,3,22.0,1,0,7.2500,True,False,True
1,1,1,38.0,1,0,71.2833,False,False,False
2,1,3,26.0,0,0,7.9250,False,False,True
3,1,1,35.0,1,0,53.1000,False,False,True
4,0,3,35.0,0,0,8.0500,True,False,True


In [20]:
# X contains the input features
# y contains the target we want to predict
X = df.drop("Survived", axis=1)
y = df["Survived"]
# Use 80% of the data for training
# Use 20% of the data for testing
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [21]:
# Create a Logistic Regression model
model = LogisticRegression(max_iter=1000)
# Train the model using the training data
model.fit(X_train, y_train)
# Predict the survival of passengers in the test dataset
predictions = model.predict(X_test)

In [22]:
# Calculate the evaluation metrics
accuracy = accuracy_score(y_test, predictions)
precision = precision_score(y_test, predictions)
recall = recall_score(y_test, predictions)
f1 = f1_score(y_test, predictions)
# Print the results
print("Accuracy :", accuracy)
print("Precision:", precision)
print("Recall   :", recall)
print("F1 Score :", f1)

Accuracy : 0.8100558659217877
Precision: 0.7857142857142857
Recall   : 0.7432432432432432
F1 Score : 0.7638888888888888


In [23]:
# Try another classification model
from sklearn.tree import DecisionTreeClassifier
# Create the Decision Tree model
tree = DecisionTreeClassifier(random_state=42)
# Train the model
tree.fit(X_train, y_train)
# Make predictions
tree_predictions = tree.predict(X_test)
# Evaluate the model
tree_accuracy = accuracy_score(y_test, tree_predictions)
tree_precision = precision_score(y_test, tree_predictions)
tree_recall = recall_score(y_test, tree_predictions)
tree_f1 = f1_score(y_test, tree_predictions)
# Print the evaluation metrics
print("Decision Tree Results")
print("----------------------")
print("Accuracy :", tree_accuracy)
print("Precision:", tree_precision)
print("Recall   :", tree_recall)
print("F1 Score :", tree_f1)

Decision Tree Results
----------------------
Accuracy : 0.7932960893854749
Precision: 0.7605633802816901
Recall   : 0.7297297297297297
F1 Score : 0.7448275862068966


## Model Performance Summary

In this project, I built two classification models, Logistic Regression and Decision Tree, using the Titanic dataset to predict passenger survival. Before training the models, the dataset was cleaned by handling missing values, removing unnecessary columns, and converting categorical variables into numerical values.

The Logistic Regression model achieved an accuracy of **81.01%**, a precision of **78.57%**, a recall of **74.32%**, and an F1-score of **76.39%**. The Decision Tree model achieved an accuracy of **79.33%**, a precision of **76.06%**, a recall of **72.97%**, and an F1-score of **74.48%**.

Overall, the Logistic Regression model performed better than the Decision Tree model across all evaluation metrics, making it the preferred model for this task. The model could be further improved by tuning its hyperparameters, adding more useful features, or trying more advanced algorithms such as Random Forest.